File: BuddhaGameServer/src/network/socket/SocketNet.java
package network.socket;

import java.util.LinkedList;
import java.util.List;
import java.util.concurrent.*;

public abstract class SocketNet {
    protected int mPort;
    protected List<SocketListener> mReceivers;


    private ThreadPoolExecutor mExecutor;

    protected SocketNet(int port) {
        mPort = port;
        mReceivers = new LinkedList<>();
        mExecutor = new ThreadPoolExecutor(
                64,
                Integer.MAX_VALUE,
                1,
                TimeUnit.MINUTES,
                new LinkedBlockingQueue<Runnable>(8),
                Executors.defaultThreadFactory(),
                new ThreadPoolExecutor.DiscardPolicy()
        );
    }

    public abstract void sendBytes(byte[] data, SocketAddress socketAddress);

    public abstract void release();

    public void addReceiver(SocketListener r){
        mReceivers.add(r);
    }

    public void removeReceiver(SocketListener r){
        mReceivers.remove(r);
    }

    protected void notify(SocketAddress address, byte[] bytes) throws Exception {
        for(SocketListener l : mReceivers){
            mExecutor.execute(() -> {
                try {
                    l.onReceive(address, bytes);
                } catch (Exception e) {
                    e.printStackTrace();
                }
            });
        }
    }

    protected void notifyError(Exception e){
        for(SocketListener l : mReceivers){
            mExecutor.execute(() -> {
                l.onError(e);
            });
        }
    }


}

File: BuddhaGameClient/Assets/Scripts/Network/Socket/SocketNet.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Threading;
using UnityEngine;

namespace BuddhaGame
{
    public abstract class SocketNet
    {
        protected String mServerIp;
        protected int mServerPort;
        protected int mLocalPort = 60000;
        protected bool mIsConnected = false;
        protected List<SocketListener> mSocketListeners;       

        protected SocketNet()
        {
            mSocketListeners = new List<SocketListener>();    
        }

        public abstract void connect(String ip, int port);

        public abstract void reconnect();

        public abstract void release();  

        public void addReceiveListener(SocketListener listener)
        {
            mSocketListeners.Add(listener);
        }

        public void removeReceiveListener(SocketListener listener)
        {
            mSocketListeners.Remove(listener);
        }

        public abstract void sendBytes(byte[] bytes);   

        protected void notifyReceive(byte[] data)
        {
            foreach (SocketListener l in mSocketListeners)
            {
                l.onReceive(data);
            }
        }

        protected void notifyError(Exception e)
        {
            foreach (SocketListener l in mSocketListeners)
            {
                l.onError(e);
            }
        }
        
    }
}
